README.md
# dev-serv

[![NPM version][npm-image]][npm-url] [![changelog][changelog-image]][changelog-url] [![license][license-image]][license-url]

[npm-image]: https://img.shields.io/npm/v/dev-serv.svg?style=flat-square
[npm-url]: https://npmjs.org/package/dev-serv
[license-image]: https://img.shields.io/badge/License-MIT-blue.svg?style=flat-square
[license-url]: https://github.com/ufologist/dev-serv/blob/master/LICENSE
[changelog-image]: https://img.shields.io/badge/CHANGE-LOG-blue.svg?style=flat-square
[changelog-url]: https://github.com/ufologist/dev-serv/blob/master/CHANGELOG.md

开发时使用的专属服务器. **后来发现通过扩展 puer 能够更好的实现(不需要开两个服务了), 因此已经转投到[puer-proxy](https://github.com/ufologist/puer-proxy)**

* 安装 `dev-serv` 模块

  `npm install dev-serv --save-dev`
* 增加服务器的配置文件
  
  从 `dev-serv/example` 中复制 `_dev-serv.json` 到项目根目录

  配置项说明请参考 [src/dev-serv.js#DEV_SERV_CONFIG](https://github.com/ufologist/dev-serv/blob/master/src/dev-serv.js), 一般仅需要配置 `backend` 参数即可
* 增加 [puer-mock](https://github.com/ufologist/puer-mock) 接口配置文件

  请参考 [puer-mock 使用文档](https://github.com/ufologist/puer-mock#usage), 复制 `node_modules/puer-mock/example` 下面的所有文件到项目根目录
* 配置项目的 `package.json` 脚本

  ```json
  "scripts": {
      "dev": "dev-serv"
  }
  ```
* 启动服务器

  `npm run dev`

## 开发时使用的服务器需要具备哪些功能

* 静态文件服务器
* 静态文件修改后浏览器自动刷新
* Mock API Server
* 代理后端真实接口(最好 Mock API Server 能够灵活切换到真实接口或者使用 Mock 接口)

即需要集 `static server + mock server + proxy server + live-reload` 于一身

## 实现方案

* 使用 expres 来做静态文件服务器                      -> 正式环境为 nginx
* 使用 puer-mock 来做 API Server(代理后端的真实接口)  -> 方便切换 mock 接口或者真实接口
* 使用 http-proxy-middleware 来代理 API Server       -> 正式环境为 nginx 反向代理(或者跨域调用)
* 使用 puer 中间件来做静态文件修改后浏览器自动刷新      -> 方便开发

即形成下面的结构

* 通过 `:8000` 端口服务静态文件
* 通过 `:8000/api` 代理在 `:8001` 端口的后端接口
  * puer-mock 在 `:8001` 端口启动 mock 接口服务
  * puer-mock 同时代理 `:18520` 端口的后端真实接口, 方便开发时可以随时切换成真实接口

```
【静态服务器】                                                    【Mock API Server】
http://localhost:8000                                            http://localhost:8001
┏━━━━━━━━━━━━━━━━━━┓                            ┏━━━━━━━━━━━━━━━━━━┓ 
┃                                 ┃                            ┃                                 ┃
┃ http://localhost:8000/a.html    ┃                            ┃ Mock API Route Config           ┃
┃ http://localhost:8000/a.css     ┃                            ┃ http://localhost:8001/user/list ┃
┃                                 ┃                            ┃                                 ┃
┃ http://localhost:8000/api       ┃ --http-proxy-middleware--> ┃      Mock API Server(puer-mock) ┃
┃                                 ┃                            ┗━━━━━━━━━━━━━━━━━━┛
┃          Static Server(express) ┃                                           
┃          + Auto Reload          ┃                                           ┃
┃       (puer-connect-middleware) ┃                                          proxy
┗━━━━━━━━━━━━━━━━━━┛                                           ┃
                                                                                v

                                                                 【后端真实接口服务器】
                                                                 http://localhost:18520
                                                                 ┏━━━━━━━━━━━━━━━━━━┓
                                                                 ┃                                 ┃
                                                                 ┃http://localhost:18520/user/list ┃
                                                                 ┃                                 ┃
                                                                 ┃             API Server(Backend) ┃
                                                                 ┗━━━━━━━━━━━━━━━━━━┛
```

因此页面上面所有接口的根路径应该为 `:8000/api`

例如:

```javascript
// 通过代理调用后端接口, 会去请求 puer-mock 的服务
// 因此实际上请求会发给 http://localhost:8001/user/list
//
// 如果 puer-mock 上禁用了这个 mock 接口, 由于 puer-mock 同时代理了后端真实接口
// 因此实际上请求会发给 http://localhost:18520/user/list
$.get('http://localhost:8000/api/user/list');
```

CHANGELOG.md
# CHANGELOG

* v0.1.1 2017-3-1

  fixbug: `http-proxy-middleware` 代理的 `context` 应该已以斜杠结尾, 否则会匹配到 `/api123` 这样的路径

* v0.1.0 2017-2-28

  实现开发时使用的专属服务器

src/dev-serv.js
/**
 * dev-server - 开发时使用的专属服务器
 */

// 开发时使用的服务器需要具备哪些功能
// * 静态文件服务器
// * 静态文件修改后浏览器自动刷新
// * Mock API Server
// * 代理后端真实接口(最好 Mock API Server 能够灵活切换到真实接口或者使用 Mock 接口)
//
// 即需要集 static server + mock server + proxy server + live-reload 于一身
//
// 实现方案
// * 使用 expres 来做静态文件服务器                      -> 正式环境为 nginx
// * 使用 puer-mock 来做 API Server(代理后端的真实接口)  -> 方便切换 mock 接口或者真实接口
// * 使用 http-proxy-middleware 来代理 API Server       -> 正式环境为 nginx 反向代理(或者跨域调用)
// * 使用 puer 中间件来做静态文件修改后浏览器自动刷新      -> 方便开发
//
// 即形成下面的结构
// --------------------
// [Static Server(express)] ==http-proxy-middleware==> Mock API Server(puer-mock) ==proxy==> API Server(Backend)
//          +
// [Auto Reload(puer-connect-middleware)]
var path = require('path');
var fs = require('fs');
var http = require('http');
var os = require('os');
var exec = require('child_process').exec;

var express = require('express');
var puer = require('puer');
// openBrowser, merge
var puerHelper = require('puer/lib/helper');
var proxy = require('http-proxy-middleware');
var serveIndex = require('serve-index');

function getDevServConfig() {
    var devServConfig = {};
    try {
        // 读取开发服务器的配置, 没有配置文件也可以
        var json = fs.readFileSync(path.join(process.cwd(), devServConfigFile));
        devServConfig = JSON.parse(json);
    } catch (error) {
        console.error('getDevServConfig: ' + error);
    }
    return devServConfig;
}

// 使用 puer-mock 来做 API Server(代理后端的真实接口)
// puer -p 8001 -a _mockserver.js -t http://localhost:18520
function startMockServer() {
    puer({
        port: devServConfig.mockServer.port,
        addon: path.join(devServConfig.staticServer.dir, devServConfig.mockServer.addon),
        proxy: devServConfig.mockServer.backend,
        launch: false
    });
}

// 使用 http-proxy-middleware 来代理 API Server
function proxyMockServer() {
    // 代理后端接口(这里通过 puer-mock 来控制是使用 mock 接口还是真实接口)
    // 当然了, 你也可以分别在不同的 route 路径上使用不同的代理
    // 例如 app.use('/api', proxy({target: 'http://localhost:18520'}); // 真实接口
    //      app.use('/mock', proxy({target: 'http://localhost:8001'}); // mock接口
    app.use(proxy(devServConfig.staticServer.apiPath, {
        target: 'http://localhost:' + devServConfig.mockServer.port,
        changeOrigin: true,
        // https://webpack.js.org/configuration/dev-server/#devserver-proxy
        // If you don't want /api to be passed along, we need to rewrite the path:
        // {'^/api/': ''}
        pathRewrite: devServConfig.staticServer.pathRewrite
    }));
}

// 使用 puer 中间件来做文件修改后浏览器自动刷新
function liveReload() {
    app.use(puer.connect(app, server, {
        dir: devServConfig.staticServer.dir
    }));
}

// 使用 expres 来做静态文件服务器
function static() {
    app.use(express.static(devServConfig.staticServer.dir));
    // Serve directory listings
    app.use(serveIndex(devServConfig.staticServer.dir, {
        icons: true
    }));
}

var app = express();
var server = http.createServer(app);
var devServConfig = {};
var devServConfigFile = '_dev-serv.json';

var DEV_SERV_CONFIG = {
    staticServer: {
        dir: process.cwd(), // 静态资源文件的根目录(默认为命令执行的目录)
        port: 8000,         // 静态服务器的端口
        apiPath: '/api/'    // 后端接口的根路径, 注意以斜杠结尾, 否则会匹配到 /api123 这样的路径
    },
    mockServer: {
        port: 8001,         // mock 服务器的端口
        addon: '_mockserver.js', // puer 的 mock 功能
        backend: 'http://localhost:18520' // puer 的 proxy 功能, 代理真实的后端接口, 方便切换
    }
}

function initDevServConfig() {
    devServConfig = puerHelper.merge(getDevServConfig(), DEV_SERV_CONFIG);
    devServConfig.staticServer.pathRewrite = {};
    devServConfig.staticServer.pathRewrite['^' + devServConfig.staticServer.apiPath] = '';
}

function startDevServer() {
    initDevServConfig();
    startMockServer();
    proxyMockServer();
    liveReload();
    static();

    server.listen(devServConfig.staticServer.port, function() {
        console.log('-------------------------------------------------');
        console.log('static server start at localhost:' + server.address().port);
        puerHelper.openBrowser('http://localhost:' + server.address().port, function(err) {
            return console.log(err || 'launch your default browser');
        });
    });
}

module.exports = {
    startDevServer: startDevServer
}